File: leetcode/894_All_Possible_Full_Binary_Trees/894_All_Possible_Full_Binary_Trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> allPossibleFBT(int N) {
        List<List<TreeNode>> rootIteration = new ArrayList<>();
        if (N > 0) {
            rootIteration.add(new ArrayList<TreeNode>());
            rootIteration.get(0).add(new TreeNode(0));
        }
        if (N > 1) rootIteration.add(new ArrayList<TreeNode>());
        for (int sz = 2; sz <= N; sz++) {
            List<TreeNode> currentPossible = new ArrayList<TreeNode>();
            for (int leftSz = 1; leftSz <= sz - 1; leftSz++) {
                List<TreeNode> leftPossible = rootIteration.get(leftSz - 1);
                List<TreeNode> rightPossible = rootIteration.get(sz - leftSz - 1);
                if (leftPossible.size() > 0 && rightPossible.size() > 0) {             
                    for (TreeNode leftRoot : leftPossible) {
                        for (TreeNode rightRoot : rightPossible) {
                            TreeNode currentRoot = new TreeNode(0);
                            currentRoot.left = leftRoot;
                            currentRoot.right = rightRoot;
                            currentPossible.add(currentRoot);
                        }
                    }
                }
            }
            rootIteration.add(currentPossible);
        }
        return rootIteration.get(N - 1);
    }
}

File: leetcode/894_All_Possible_Full_Binary_Trees/894_All_Possible_Full_Binary_Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> allPossibleFBT(int N) {
        vector<TreeNode*> res = vector<TreeNode*>();
        if (N % 2 == 0) return res;
        if (N == 1) {
            res.push_back(new TreeNode(0));
            return res;
        } 
        for (int leftSz = 1; leftSz <= N - 2; leftSz++) {
            vector<TreeNode*> leftPos = allPossibleFBT(leftSz);
            vector<TreeNode*> rightPos = allPossibleFBT(N - 1 - leftSz);
            if (leftPos.size() > 0 && rightPos.size() > 0) {
                for (TreeNode* leftNode : leftPos) {
                    for (TreeNode* rightNode : rightPos) {
                        TreeNode* root = new TreeNode(0);
                        root->left = leftNode;
                        root->right = rightNode;
                        res.push_back(root);
                    }
                }
            }
        }
        return res;
    }
};
